src/Language/PureScript/TypeChecker.hs

```diff
@@ -434,31 +434,93 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
   warnAndRethrow (addHint (ErrorInModule mn)) $ do
     modify (\s -> s { checkCurrentModule = Just mn })
     decls' <- typeCheckAll mn exps decls
+    classesToSuperClasses <- gets
+      ( M.map 
+        ( S.fromList 
+        . fmap constraintClass 
+        . typeClassSuperclasses
+        )
+      . typeClasses
+      . checkEnv
+      )
+    let 
+      -- A mapping of class names to their declaration refs in this module.
+      moduleClassDeclarations :: M.Map (Qualified (ProperName 'ClassName)) DeclarationRef
+      moduleClassDeclarations = 
+        M.fromList
+          . mapMaybe (\x -> case x of
+                td@(TypeClassDeclaration _ name _ _ _ _) ->
+                  (,) (Qualified (Just mn) name) <$> classDecToRef td
+                _ ->
+                  Nothing)
+          $ decls
+      classDecToRef :: Declaration -> Maybe DeclarationRef
+      classDecToRef (TypeClassDeclaration (ss', _) n _ _ _ _) = Just (TypeClassRef ss' n)
+      classDecToRef _ = Nothing
+
+      moduleClassDeclarationSet :: S.Set (Qualified (ProperName 'ClassName))
+      moduleClassDeclarationSet = 
+        M.foldrWithKey (\k _ -> S.insert k) S.empty moduleClassDeclarations
+
+      -- A function that, given a class name, returns the set of
+      -- transitive class dependencies that are defined in this
+      -- module.
+      transitiveSuperClassesFor
+          :: Qualified (ProperName 'ClassName)
+          -> S.Set (Qualified (ProperName 'ClassName))
+      transitiveSuperClassesFor qname =
+        S.intersection moduleClassDeclarationSet
+          $ untilSame 
+            (\s -> s <> foldMap (\n -> fromMaybe S.empty (M.lookup n classesToSuperClasses)) s)
+            (fromMaybe S.empty (M.lookup qname classesToSuperClasses))
```

Comment:
This is going to do a lot of recomputation to find the fixed point of that set.

Reply:
I suppose you could argue that it's not necessary to obtain transitive superclasses, only immediate ones, as I think this check will fail in the transitive case if and only if it fails in the immediate case. Obtaining transitive dependencies just means we get a more useful error if it does fail, I think?

E.g. If we're exporting `C3` here, we also need to export `C2` and `C1`:

```purescript
module X (class C3) where
class C1 a
class (C1 a) <= C2 a
class (C2 a) <= C3 a
```

And whether we obtain transitive dependencies here or not, we're going to fail with a transitive export error; obtaining transitive superclasses presumably just means that we have a more useful error, which lists both `C2` and `C1` as opposed to only `C2`.

Perhaps as a middle ground we could only obtain immediate superclasses in the first pass, and then if there are any errors, obtain transitive superclasses for the offending classes for the sake of a more detailed error? Does that sound feasible?

Reply:
That is a really good point -- we're already checking every exported item in the list, so we only need to verify that the immediate superclasses (also defined in the module) are exported. The transitive superclasses defined in module would be caught on when they're being inspected :smile: 

You're correct that this method does provide a nicer error message, in that you'll get one error message telling you to add all the things to the exports.

We can get "best of both worlds" if we can catch the error and add the lookup for transitive superclasses there before rethrowing.